tests/unit_tests/test_asyncio_extensions.py

```diff
@@ -0,0 +1,185 @@
+
+
+@fast_forward(500)
+async def test_synchronized():
+    in_call = False
+
+    @synchronized
+    async def sleep_for_1s():
+        nonlocal in_call
+
+        assert in_call is False, "Multiple concurrent executions!"
+
+        in_call = True
+        await asyncio.sleep(1)
+        in_call = False
+
+    # 500 calls * 1 second each should sleep for 500 seconds
+    await asyncio.gather(*[sleep_for_1s() for _ in range(500)])
+
+
+@fast_forward(500)
+async def test_synchronized_empty_args():
+    in_call = False
+
+    @synchronized()
+    async def sleep_for_1s():
+        nonlocal in_call
+
+        assert in_call is False, "Multiple concurrent executions!"
+
+        in_call = True
+        await asyncio.sleep(1)
+        in_call = False
+
+    # 500 calls * 1 second each should sleep for 500 seconds
+    await asyncio.gather(*[sleep_for_1s() for _ in range(500)])
+
+
+@fast_forward(500)
+async def test_synchronized_with_lock():
+    lock = asyncio.Lock()
+    in_call = False
+
+    @synchronized(lock)
+    async def sleep_for_1s():
+        nonlocal in_call
+
+        assert lock.locked(), "Lock was not held during function execution!"
+        assert in_call is False, "Multiple concurrent executions!"
+
+        in_call = True
+        await asyncio.sleep(1)
+        in_call = False
+
+    # 500 calls * 1 second each should sleep for 500 seconds
+    await asyncio.gather(*[sleep_for_1s() for _ in range(500)])
+
+
+@fast_forward(500)
+async def test_synchronizedmethod():
+    class Test(object):
+        def __init__(self):
+            self.in_call = False
+
+        @synchronizedmethod
+        async def sleep_for_1s(self):
+            assert self.in_call is False, "Multiple concurrent executions!"
+
+            self.in_call = True
+            await asyncio.sleep(1)
+            self.in_call = False
+
+    a = Test()
+    b = Test()
+    # Calls to different instances should not block eachother
+    await asyncio.wait(
+        [a.sleep_for_1s() for _ in range(500)] +
+        [b.sleep_for_1s() for _ in range(500)],
+        timeout=502
+    )
+
+
+@fast_forward(500)
+async def test_synchronizedmethod_attrname():
+    class Test(object):
+        def __init__(self):
+            self.in_call = False
+            self._my_lock = asyncio.Lock()
+
+        @synchronizedmethod("_my_lock")
+        async def sleep_for_1s(self):
+            assert self._my_lock.locked(), \
+                "Lock was not held during function execution!"
+            assert self.in_call is False, "Multiple concurrent executions!"
+
+            self.in_call = True
+            await asyncio.sleep(1)
+            self.in_call = False
+
+    a = Test()
+    b = Test()
+    # Calls to different instances should not block eachother
+    await asyncio.wait(
+        [a.sleep_for_1s() for _ in range(500)] +
+        [b.sleep_for_1s() for _ in range(500)],
+        timeout=502
+    )
+
+
+@fast_forward(500)
+async def test_synchronizedmethod_class_invocation():
```

Comment:
Is this not identical to `test_synchronizedmethod` since the interpreter parses `a.sleep_for_1s()` as `Test.sleep_for_1s(a)` anyways? Or is there a way to make one of these tests pass but not the other?

Reply:
Riiight. This can probably be deleted. This was necessary when I was still manually implementing a descriptor (object with `__get__` method) because I needed to verify that my descriptor implementation would correctly for both invocations. But then I realized I didn't need to use a custom descriptor at all, and went back to just using normal functions.

Still, I learned a lot about how python objects actually work under the hood. In fact the calls `a.foo()` and `A.foo(a)` don't do *exactly* the same thing. For instance if you fire up a python interpreter you can easily check that `A.foo is A.foo` is true, but `a.foo is a.foo` is not. I found this documentation quite helpful https://docs.python.org/3/howto/descriptor.html

Reply:
Thanks for the link!